autosubmit_api/experiment/utils.py

```diff
@@ -20,4 +22,56 @@ def get_cas_user_from_xml(xmlstring):
     except Exception as exp:
       print(exp)
       print((traceback.format_exc))
-    return user
\ No newline at end of file
+    return user
+
+def read_tail(file_path: str, num_lines: int =150) -> List[dict]:
+    """
+    Reads the last num_lines of a file and returns them as a dictionary.
+   
+    :param file_path: path to the file
+    :param num_lines: number of lines to read
+    """
+    tail_content = []
+
+    process = subprocess.Popen(
+        ["tail",  "-{0}".format(num_lines), file_path],
+        stdout=subprocess.PIPE,
+        stderr=subprocess.PIPE,
+    )
+    
+    stdout, stderr = process.communicate()
+
+    if process.returncode != 0:
+        raise ValueError(stderr.decode())
+    
+    lines = stdout.decode().splitlines()
+    for i, item in enumerate(lines):
+        tail_content.append({"index": i, "content": item})
+  
+    return tail_content
+
+def get_files_from_dir_with_pattern(dir_path: str, pattern: str) -> List[str]:
+  """
+  Returns a list of files ordered by creation date in a directory that match a pattern.
+  """
+  # Submits the commands
+  ls_process = subprocess.Popen(
+    ['ls', '-t', dir_path],
+    stdout=subprocess.PIPE
+  )
+  grep_process = subprocess.Popen(
+    ['grep', pattern],
+    stdin=ls_process.stdout,
+    stdout=subprocess.PIPE,
+    stderr=subprocess.PIPE
+  )
+
+  # Allow ls_process to receive a SIGPIPE if grep_process exits
+  ls_process.stdout.close()
```

Comment:
I didn't understand why .close()` was called here. Reading the pydoc comment, "Returns a list of files ordered by creation date in a directory that match a pattern.", here's how I'd do that (`ls -t` seems to bring them in desc order):

```py
# previously created `/tmp/1` dir, with files a, z, b, c (in this order)
import pathlib

dir_path = '/tmp/1'
pattern = '**/*.*'

def get_files_from_dir_with_pattern(dir_path: str, pattern: str) -> List[str]:
    """..."""
    return [sorted(pathlib.Path(dir_path).glob(pattern), key=lambda f: f.stat().st_mtime, reverse=True)]
```

I think that way we don't need to fork new processes... but not sure if the pattern is equivalent -- would need to be tested/checked.

Reply:
Good one! I was focusing so much on just updating the `popen` that I didn't think about not using it at all which is a better solution for improving compatibility.

Reply:
Commit https://github.com/BSC-ES/autosubmit-api/pull/150/commits/e3bef952b97b50ec18718563a04553f7a8363309 uses your nice snippet. I added a test just to be sure it works the same before and after the modification :nerd_face: